From c2dfc902447d937ff85e2c3649ad004d99aae85d Mon Sep 17 00:00:00 2001
From: Anas Nashif <anas.nashif@intel.com>
Date: Mon, 19 Dec 2011 11:31:47 +0000
Subject: [PATCH 5/6] client loop

---
 clientloop.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/clientloop.c b/clientloop.c
index de79793..4f5ebbc 100644
--- a/clientloop.c
+++ b/clientloop.c
@@ -1933,7 +1933,7 @@ client_input_channel_req(int type, u_int32_t seq, void *ctxt)
 		}
 		packet_check_eom();
 	}
-	if (reply) {
+	if (reply && c != NULL) {
 		packet_start(success ?
 		    SSH2_MSG_CHANNEL_SUCCESS : SSH2_MSG_CHANNEL_FAILURE);
 		packet_put_int(c->remote_id);
-- 
1.7.6.4

